File: app_java/src/main/java/vlab/android/architecture/di/module/AppModule.java
package vlab.android.architecture.di.module;


import android.app.Application;
import android.content.Context;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

/**
 * Created by vinhtran on 2/11/18.
 */

@Module(
        includes = {
                ViewModelModule.class,
                UiModule.class,
                NetworkModule.class,
                ApiModule.class,
                RepositoryModule.class
        }
)
public class AppModule {

    private Application mApplication;
    private Context mApplicationContext;

    public AppModule(Application application) {
        this.mApplication = application;
        this.mApplicationContext = application.getApplicationContext();
    }

    @Provides
    @Singleton
    public Application provideApplication() {
        return mApplication;
    }

    @Provides
    @Singleton
    Context provideApplicationContext() {
        return mApplicationContext;
    }

}

File: app_kotlin/src/main/java/vlab/android/architecture/di/AppModule.kt
package vlab.android.architecture.di


import android.app.Application
import android.content.Context
import dagger.Module
import dagger.Provides
import vlab.android.architecture.di.module.ApiModule
import vlab.android.architecture.di.module.NetworkModule
import vlab.android.architecture.di.module.RepositoryModule
import vlab.android.architecture.di.module.UiModule
import vlab.android.architecture.di.module.ViewModelModule
import javax.inject.Singleton

/**
 * Created by vinhtran on 2/11/18.
 */

@Module(includes = arrayOf(
        ViewModelModule::class,
        UiModule::class,
        NetworkModule::class,
        ApiModule::class,
        RepositoryModule::class)
)
class AppModule {

    private val mApplication: Application

    constructor(mApplication: Application) {
        this.mApplication = mApplication
        this.mApplicationContext = mApplication.applicationContext
    }

    private val mApplicationContext: Context

    @Provides
    @Singleton
    fun provideApplication(): Application {
        return mApplication
    }

    @Provides
    @Singleton
    internal fun provideApplicationContext(): Context {
        return mApplicationContext
    }

}
